### ISSUE TITLE
Should IsMatchInBypassList should accept localhost has any IP Adress loopback ?

### ISSUE THREAD

**Issue by Davilink** [dotnet/runtime#31680]
After a few hours of investigation of why my request was proxified, i manage to find that the reason was because in my NO_PROXY variable i had only "localhost", but the request host was 127.0.0.1
so i change my NO_PROXY variable to "localhost,127.0.0.1" and the request stopped to be proxyfied.

https://github.com/dotnet/runtime/blob/60f6129b3eaef088f4b1ddc30d3320a68f396d85/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L231

**Comment by scalablecory** [comment]
Does this behavior differ from how wininet/winhttp uses the wininet exception list?

Note that "localhost" can technically be mapped to a non-loopback address, though that would be weird.

**Comment by stephentoub** [comment]
I agree with @scalablecory.  This concerns me in two ways.  First, anything can be mapped to 127.0.0.1, so even if we special-case "localhost", we wouldn't be catching any other such mapping.  Second, "localhost" can be mapped to other addresses instead, and in fact some tools do that as part of simulating local access while actually doing a form of remote debugging; it seems potentially problematic for the implementation to just always assume they're related.

cc: @wfurt

**Comment by Davilink** [comment]
If i want to "NO_PROXY" every loopback address because i use an script that
will dynamicly bind to any loopback ip address how will i proceed to do it?

Le ven. 7 févr. 2020 19 h 59, Stephen Toub <notifications@github.com> a
écrit :

**Comment by Davilink** [comment]
NO_PROXY is suppose to contains a list of host name, "localhost" is reserved top domain hostname, but 127.0.0.1 - 127.255.255.255 are not host name, they are IP Adresses.

according to https://tools.ietf.org/html/rfc2606#section-2
` The ".localhost" TLD has traditionally been statically defined in
      host DNS implementations as having an A record pointing to the
      loop back IP address and is reserved for such use.  Any other use
      would conflict with widely deployed code which assumes this use.
`
it seem acceptable to consider that "localhost" should be interpreted has loopback ip addresses, but maybe we could use "*.localhost" instead

```csharp
private bool IsMatchInBypassList(Uri input)
{
    if (_bypass != null)
    {
        foreach (string s in _bypass)
        {
            if (s[0] == '.')
            {
                // This should match either domain it self or any subdomain or host
                // .foo.com will match foo.com it self or *.foo.com
                if ((s.Length - 1) == input.Host.Length &&
                    string.Compare(s, 1, input.Host, 0, input.Host.Length, StringComparison.OrdinalIgnoreCase) == 0)
                {
                    return true;
                }
                else if (input.Host.EndsWith(s, StringComparison.OrdinalIgnoreCase))
                {
                    return true;
                }

            }
            else if (string.Equals(s, input.Host, StringComparison.OrdinalIgnoreCase))
            {
                return true;
            }
            else if (string.Equals(s, "*.localhost", StringComparison.OrdinalIgnoreCase))
            {
                if(IPAddress.TryParse(input.Host, out var ipAddress))
                {
                    return IPAddress.IsLoopback(ipAddress);
                }
            }
        }
    }
    return false;
}
```

**Comment by wfurt** [comment]
I'm wondering if there is an opportunity to improve diagnostic. It would be nice if it is easy to diagnose issues like this. 

It seems strange that for "loopback" we would try to find all addresses but we would not do it for other names.

**Comment by Davilink** [comment]
I find (by googling) this page, maybe we could take some inspiration : 
https://developer.mozilla.org/en-US/docs/Archive/Misc_top_level/No_Proxy_For_configuration
Like allowing the use of ip address with CIDR to support range of ip adress in the NO_PROXY

**Comment by karelz** [comment]
Triage: We should find out what WinHttp and curl do and probably try to match that behavior.

@Davilink is it something you can help us investigate please?

**Comment by Davilink** [comment]
@karelz i will try

**Comment by Davilink** [comment]
According to the source code of curl, the no_proxy doesn't support IP address whitelist by CIDR neither interpret localhost as go to no proxy all the loopback ip address
https://github.com/curl/curl/blob/1d1e9e8ad7a5f35818dc7072e6fbe8a76a93e018/lib/url.c#L2044-L2116

**Comment by slaneyrw** [comment]
Ruby and Go both support CIDR blocks in their no_proxy settings, so parts of the industry are already ahead of the standards in this regard.

This is become more of an issue since the deprecation of HttpClientHandler ( and the DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLER env variable ).  A lot of corporate env use proxy servers and VPNs, a simple IP Address is not sufficient to cover case.

**Comment by slaneyrw** [comment]
> I find (by googling) this page, maybe some inspiration can be taken: https://developer.mozilla.org/en-US/docs/Archive/Misc_top_level/No_Proxy_For_configuration Like allowing the use of ip address with CIDR to support range of ip adress in the NO_PROXY

@wfurt Responding to this
